[processor/transform] Create `With*Functions` factory options to provide custom OTTL funcs programatically.

franciscovalentecastro opened the pull request:
#### Description
Create `With*Functions` factory options to provide custom OTTL functions for logs, metrics or traces to the resulting processor. Also created the "exported" `Default*Functions` functions that can be used to provide a "subset" of the default functions used in the the transform processor. 

Since in the current collector design the `Config` doesn't have access to the `Factory`, the `Config.Validate` method can't access the additional OTTL functions set in the `NewFactory`. I have found two ways to circumvent this :
 
  - **[Option 1, Current]** Add un-exported fields of the form `*Functions` to the `Config` struct to use in the `Validate()` method.
  - **[Option 2]** Create `ValidateFunctions` methods that can be used to create a custom `Config.Validate()`.

Some details about the solution :

- To create a custom `transform` processor the user needs to create a `NewFactory` providing the custom OTTL functions.
- This is a result of the discussion in this thread : https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39641#discussion_r2081450587
- This is intended for developers of Opentelemetry Collector Distributions.

#### Usage

The factory options can be used like the following example :
```
NewFactoryWithOptions(
	WithLogFunctions(transformprocessor.DefaultLogFunctions()),
	WithLogFunctions([]ottl.Factory[ottllog.TransformContext]{FunctionOne, FunctionTwo, ...}),
	WithLogFunctions([]ottl.Factory[ottllog.TransformContext]{FunctionThree, FunctionFour, ...}),
)
```


<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/39698

**Note : There is a similar followup proposed update for the `processor/filter`.**

<!--Describe what testing was performed and which tests were added.-->
#### Testing
- Added unit tests for the modified methods and functions.

<!--Describe the documentation added.-->
#### Documentation
- Added comments to the new `With*Functions`, `Default*Functions`  and `FactoryWithOptions` functions.

<!--Please delete paragraphs that you did not use before submitting.-->

franciscovalentecastro committed:
[processor/transform] Create `additional*Funcs` parameters for NewProcessor and *Functions methods.

franciscovalentecastro committed:
[processor/transform] Create `transformProcessorFactory` and `WithAdditional*Functions` factory options.

franciscovalentecastro committed:
[processor/transform] Create `Validate*Statements` methods to be used for validation of custom transform processor distributions.

franciscovalentecastro committed:
[processor/transform] Update README.md.

quentinmit commented on processor/transformprocessor/factory.go:
```diff
@@ -22,13 +25,67 @@ import (
+// FactoryOption applies changes to transformProcessorFactory.
+type FactoryOption func(factory *transformProcessorFactory)
+
+// defaultAdditionalMetricFunctions returns an empty slice of ottl function factories.
+func defaultAdditionalMetricFunctions() []ottl.Factory[ottlmetric.TransformContext] {
+	return []ottl.Factory[ottlmetric.TransformContext]{}
+}
+
+// defaultAdditionalLogFunctions returns an empty slice of ottl function factories.
+func defaultAdditionalLogFunctions() []ottl.Factory[ottllog.TransformContext] {
+	return []ottl.Factory[ottllog.TransformContext]{}
+}
+
+// defaultAdditionalSpanFunctions returns an empty slice of ottl function factories.
+func defaultAdditionalSpanFunctions() []ottl.Factory[ottlspan.TransformContext] {
+	return []ottl.Factory[ottlspan.TransformContext]{}
+}
```
This is equivalent to the zero value for these fields, so there's no need for these functions to exist.

franciscovalentecastro replied:
Done. I removed these methods.

quentinmit commented on processor/transformprocessor/README.md:
```diff
@@ -726,6 +726,9 @@ transform:
 The configuration can be used also with `delete_matching_keys()` to copy the attributes that do not match the regular expression.
 
+## Additional OTTL Functions
+The Factory Options `WithAdditionalMetricFunctions`, `WithAdditionalSpanFunctions` and `WithAdditionalLogFunctions` can be set in the transform processor `NewFactory` to add additional OTTL functions to the resulting processor. The `Config` methods `ValidateMetricStatements`, `ValidateSpanStatements` and `ValidateLogStatements` to receive the additional OTTL functions to validate the context stat ...
```
`s/set in/set when calling/`

I don't think there's any need to document the change to the `Validate*` methods in the README.

What is the "custom `Config`" you're referring to?

franciscovalentecastro replied:
> What is the "custom Config" you're referring to?

My initial approach consisted in having to define a `NewFactory` with the additional functions and a new `Config.Validate` method using `ValidateWithAdditionalFunctions` to add this functions to the `Validate()` methods.

edmocosta replied:
I wouldn't add any API docs here actually, the README is more like an user-facing docs, and normally does not contains API docs. We could document it at the code level instead, using the standard godocs comments.

franciscovalentecastro replied:
Done! I removed the `README` update and i'll make sure the `FactoryOption` and `NewFactory` have godoc comments.

quentinmit commented on processor/transformprocessor/config.go:
```diff
@@ -129,45 +132,72 @@ func (c *Config) Unmarshal(conf *confmap.Conf) error {
 
 	if len(c.TraceStatements) > 0 {
-		pc, err := common.NewTraceParserCollection(component.TelemetrySettings{Logger: zap.NewNop()}, common.WithSpanParser(traces.SpanFunctions()), common.WithSpanEventParser(traces.SpanEventFunctions()))
+		err := c.ValidateTraceStatements()
```
Don't you need to pass the new functions as arguments to the new methods you've defined here?

quentinmit commented on processor/transformprocessor/config.go:
```diff
@@ -129,45 +132,72 @@ func (c *Config) Unmarshal(conf *confmap.Conf) error {
 
+func (c *Config) ValidateTraceStatements(additionalSpanFuncs ...ottl.Factory[ottlspan.TransformContext]) error {
+	var errors error
+	pc, err := common.NewTraceParserCollection(component.TelemetrySettings{Logger: zap.NewNop()}, common.WithSpanParser(traces.SpanFunctions(additionalSpanFuncs)), common.WithSpanEventParser(traces.SpanEventFunctions()))
```
Aside from the first line, these functions are all iterating over a `common.ContextStatements` object and doing the same thing. You can refactor this into a single shared helper function.

franciscovalentecastro replied:
My initial goal while creating `Validate*Statements` was to encapsulate all reference of elements from `internal/common`, `internal/metrics`, etc, into helper functions that a developer of a custom `trasform` processor could use to create a custom `Config.Validate()`.

I'm not sure if that would work with the proposed refactoring to reduce repeated code. I removed this methods in my last iteration.

quentinmit commented on processor/transformprocessor/factory.go:
```diff
@@ -22,13 +25,67 @@ import (
+	return []ottl.Factory[ottlspan.TransformContext]{}
+}
+
+// WithAdditionalMetricFunctions adds ottl metric functions to resulting processor
+func WithAdditionalMetricFunctions(metricFunctions []ottl.Factory[ottlmetric.TransformContext]) FactoryOption {
+	return func(factory *transformProcessorFactory) {
+		factory.additionalMetricFunctions = metricFunctions
+	}
+}
+
+// WithAdditionalLogFunctions adds ottl log functions to resulting processor
+func WithAdditionalLogFunctions(logFunctions []ottl.Factory[ottllog.TransformContext]) FactoryOption {
+	return func(factory *transformProcessorFactory) {
+		factory.additionalLogFunctions = logFunctions
+	}
+}
+
+// WithAdditionalSpanFunctions adds ottl span functions to resulting processor
+func WithAdditionalSpanFunctions(spanFunctions []ottl.Factory[ottlspan.TransformContext]) FactoryOption {
+	return func(factory *transformProcessorFactory) {
+		factory.additionalSpanFunctions = spanFunctions
+	}
+}
```
These options should probably live in `pkg/ottl` so they can be shared by multiple processors, instead of having to be duplicated in each processor that uses OTTL.

franciscovalentecastro replied:
I removed `pkg/ottl/ottlprocessor` based on the last response from @edmocosta regarding the complexity of creating a shared API that would work for all OTTL based components.

> Yes, I agree we should aim for consistent APIs, but I still don't think it should be enforced/defined by the OTTL package. It already provides the parsers/utilities that components need to use the language, but how they do that, is up to them.

> There are many different use-cases for OTTL at the moment (processor, connectors...), and it would be tricky to come up with some common options that perfectly fit all components. For example, providing the WithAdditionalSpanEventFunctions option for a processor that only works with logs wouldn't make much sense.

franciscovalentecastro committed:
[processor/transform] Create `ValidateWithAdditionalFunctions`  and fix `functions_test.go`.

franciscovalentecastro committed:
[processor/transform] Create `OttlProcessorFactory` and use unexported `Config` fields to set additional ottl funcs.

franciscovalentecastro commented:
Following up on the review request CC @edmocosta @TylerHelmuth @evan-bradley. This PR implements the approach discussed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39641#discussion_r2081450587.

evan-bradley reviewed (commented):
Thanks for taking this on, and sorry for the slow review.

Could we adjust the approach slightly?
1. Make `LogFunctions`, `SpanFunctions`, etc. exported at the top level of the transform processor module.
2. Instead of supplying additional functions, can we make it so the user specifies all functions they would like to include?

So the invocation would look like this:

```go
NewFactory(
	WithSpanFunctions(
		transformprocessor.DefaultSpanFunctions(),
		[]ottl.Factory[ottlspan.TransformContext]{FunctionOne, FunctionTwo, ...},
		[]ottl.Factory[ottlspan.TransformContext]{FunctionThree, FunctionFour, ...},
	),
)
```

I think this has two advantages:
1. Treats all the functions the same. The OTTL stdlib functions aren't anything special, so we shouldn't segment them in the code.
2. Allows the component author to make it so only the functions they want are enabled in the transform processor. This could be useful if, for example, a distro author wants to remove `MD5` for security reasons or standardize all statements to use the pattern replacement functions instead of the match functions.

edmocosta commented:
> Thanks for taking this on, and sorry for the slow review.
> 
> Could we adjust the approach slightly?
> 
> 1. Make `LogFunctions`, `SpanFunctions`, etc. exported at the top level of the transform processor module.
> 2. Instead of supplying additional functions, can we make it so the user specifies all functions they would like to include?
> 
> So the invocation would look like this:
> 
> ```go
> NewFactory(
> 	WithSpanFunctions(
> 		transformprocessor.DefaultSpanFunctions(),
> 		[]ottl.Factory[ottlspan.TransformContext]{FunctionOne, FunctionTwo, ...},
> 		[]ottl.Factory[ottlspan.TransformContext]{FunctionThree, FunctionFour, ...},
> 	),
> )
> ```
> 
> I think this has two advantages:
> 
> 1. Treats all the functions the same. The OTTL stdlib functions aren't anything special, so we shouldn't segment them in the code.
> 2. Allows the component author to make it so only the functions they want are enabled in the transform processor. This could be useful if, for example, a distro author wants to remove `MD5` for security reasons or standardize all statements to use the pattern replacement functions instead of the match functions.

I agree with @evan-bradley approach, and in addition to that I'd:
- Move the `pkg/ottlprocessor`  back into the transform processor package, I don't think we should have such shared options at the OTTL package, which already exports all necessary utilities to delegate this kind of logic to components (which might differ from each other).
- Add the following logs:
   - When custom functions are being used, so we know we're not dealing with the standard's transform processor behavior and extra functions might exists.
   - When an OTTL standard function is overridden. We can't support custom functions overrides upstream, and that let users and community know that it's a custom function override, and that its behavior might be different from the standard version. 
   - When custom functions with duplicate names are configured (should we allow it at all?)

quentinmit commented:
> > 1. Treats all the functions the same. The OTTL stdlib functions aren't anything special, so we shouldn't segment them in the code.
> > 2. Allows the component author to make it so only the functions they want are enabled in the transform processor. This could be useful if, for example, a distro author wants to remove `MD5` for security reasons or standardize all statements to use the pattern replacement functions instead of the match functions.

I thought about doing this, but I considered that we are likely to have other ways for users to add OTTL functions in the future (e.g. an "OTTL function" extension type, or functions loaded from config with wasm), and if we have this option pass a complete function list, it feels like that would have unclear semantics if combined with user-defined functions.

> I agree with @evan-bradley approach, and in addition to that I'd:
> 
> * Move the `pkg/ottlprocessor`  back into the transform processor package, I don't think we should have such shared options at the OTTL package, which already exports all necessary utilities to delegate this kind of logic to components (which might differ from each other).

Maybe a compromise to avoid duplicate code would be to put the options in an internal package and reexpose them in `transformprocessor`  and `filterprocessor`? I do think there's value in having a consistent API that can be used by any component with user-configured OTTL expressions.

> * Add the following logs:
>   
>   * When custom functions are being used, so we know we're not dealing with the standard's transform processor behavior and extra functions might exists.
>   * When an OTTL standard function is overridden. We can't support custom functions overrides upstream, and that let users and community know that it's a custom function override, and that its behavior might be different from the standard version.
>   * When custom functions with duplicate names are configured (should we allow it at all?)

How would you combine this with @evan-bradley's request to pass in the complete function list? At that point every function would be "custom".

edmocosta commented:
> Maybe a compromise to avoid duplicate code would be to put the options in an internal package and reexpose them in transformprocessor and filterprocessor? I do think there's value in having a consistent API that can be used by any component with user-configured OTTL expressions.

Yes, I agree we should aim for consistent APIs, but I still don't think it should be enforced/defined by the OTTL package. It already provides the parsers/utilities that components need to use the language, but how they do that, is up to them. 

There are many different use-cases for OTTL at the moment (processor, connectors...), and it would be tricky to come up with some common options that perfectly fit all components. For example, providing the `WithAdditionalSpanEventFunctions` option for a processor that only works with logs wouldn't make much sense.

> How would you combine this with @evan-bradley's request to pass in the complete function list? At that point every function would be "custom".

That's a good question, I wish we could find a good solution for that. If I had to choose between both, I would rather start limiting it to additional functions than not having  those logs. 

An option would be keeping functions segmented (`WithAdditional*Funcs`), and later - if needed - add another factory option to customize the standard functions in a controlled way.

franciscovalentecastro committed:
[processor/transform] Add more `functions` and `processor` tests.

franciscovalentecastro committed:
[processor/transform] Remove `README.md` update.

franciscovalentecastro committed:
[processor/transform] Remove `pkg/ottl/ottlprocessor` package and move `FactoryOption`s to `transformprocessor`.

franciscovalentecastro committed:
[processor/transform] Refactor `WithAdditional*Functions` to get logs from `create*Processor` when setting custom functions.

franciscovalentecastro committed:
[processor/transform] Revert changes on `internal/functions.go`.

evan-bradley commented:
>I considered that we are likely to have other ways for users to add OTTL functions in the future (e.g. an "OTTL function" extension type, or functions loaded from config with wasm)

@quentinmit That's a good consideration. Personally, I don't want to plan too far in advance for that since those ideas may not end up happening or may end up taking a different form than we'd expect. Additionally, I would personally be okay with `WithSpanFunctions` for standard Go functions, and e.g. `WithExtensionFunctions` and `WithWASMFunctions` for things that handle those two situations, for example. I think it's likely possible that we won't necessarily segment the functions by signal at that layer since the interfaces involved with supplying functions to the transform processor would probably be in a better position to separate functions by signal. This is a digression since these are both still a long ways out, but I think this illustrates why it's difficult to plan for what these might look like. I think we'll be able to make sure the API feels clean once we get to that point.

>There are many different use-cases for OTTL at the moment (processor, connectors...), and it would be tricky to come up with some common options that perfectly fit all components. For example, providing the WithAdditionalSpanEventFunctions option for a processor that only works with logs wouldn't make much sense.

@edmocosta I think we could likely define these in a common way eventually, but for now I agree it's okay for these options to live in the transform processor. We can separate them into a different package down the road once we have a better understanding of how they're used.

>That's a good question, I wish we could find a good solution for that. If I had to choose between both, I would rather start limiting it to additional functions than not having those logs.
>
>Another option would be keeping functions segmented (WithAdditional*Funcs), and later - if needed - add another factory option to customize the standard functions in a controlled way.

I think we could simply emit a log if any of the `With*Funcs` are passed to `NewFactory`: at the point the user is using these, they're at least intending to operate outside what's supported upstream. I would like to resist printing anything other than a debug log until we see that this is an issue though. I see the value in alerting the user that they should be mindful that there are customizations, but emitting a log every time the Collector starts up feels too noisy unless it's necessary.

Overall I don't think segmenting the core/user-defined functions in the `With*Funcs` options would result in significantly worse UX, but I feel like it would be a little cleaner if we treat all functions the same way.

edmocosta commented:
> I think we could simply emit a log if any of the With*Funcs are passed to NewFactory: at the point the user is using these, they're at least intending to operate outside what's supported upstream. I would like to resist printing anything other than a debug log until we see that this is an issue though. I see the value in alerting the user that they should be mindful that there are customizations, but emitting a log every time the Collector starts up feels too noisy unless it's necessary.

Agreed, those extra logs that I mentioned would be useful for troubleshooting, so I think they should be debug logs. Custom collector distributions using this new API probably wouldn't want them to be printed in any other level as well.

franciscovalentecastro committed:
[processor/transform] Create `With*Functions` factory options instead of `WithAdditional*Functions`. Create debug log.

franciscovalentecastro committed:
[processor/test] Use variadic arguments in With*Functions.

franciscovalentecastro commented:
I implemented all details and comments on this PR. A summary of the updates :

- Created `With*Functions` instead of `WithAdditional*Functions`.
- Implemented a "Debug" level in the "create processor" step when the "default" functions are overriden with 
`With*Functions`.  There is one log for each "logs, metrics or traces".
- Created `functions.go` and `functions_test.go` to define "exported" `Default*Functions` and helper functions.
- Modified the signature of the `NewProcessor` functions to receive the set custom functions and updated tests accordingly.

franciscovalentecastro committed:
[processor/transform] Add CHANGELOG entry.

franciscovalentecastro committed:
[processor/transform] Fix lint error.

franciscovalentecastro committed:
[processor/transform] Create `NewFactoryWithOptions` to avoid failure from `make checkapi`.

franciscovalentecastro committed:
[processor/transform] Add `assertConfigContainsDefaultFunctions` to assert unexported properties of config.

franciscovalentecastro committed:
[processor/transform] Update non-default functions log and remove `Sugar`.

franciscovalentecastro committed:
[processor/transform] Update `With*Functions` to make possible to use multiple times.

franciscovalentecastro commented on processor/transformprocessor/factory.go:
```diff
@@ -22,34 +27,118 @@ import (
-
-	proc, err := logs.NewProcessor(oCfg.LogStatements, oCfg.ErrorMode, oCfg.FlattenData, set.TelemetrySettings)
+	if f.defaultLogFunctionsOverriden {
+		set.Logger.Sugar().Debug("non-default ottl log functions are set in \"transform\" processor")
```
Done! Removed `Sugar`. Thanks for mentioning! At some point my linter was complaining for not using `Sugar` not sure why, but now it worked 🤷 .

franciscovalentecastro commented on processor/transformprocessor/factory.go:
```diff
@@ -22,34 +27,114 @@ import (
+// WithDataPointFunctions set ottl datapoint functions in resulting processor.
+func WithDataPointFunctions(dataPointFunctions ...[]ottl.Factory[ottldatapoint.TransformContext]) FactoryOption {
+	return func(factory *transformProcessorFactory) {
+		factory.dataPointFunctions = createFunctionsMap(dataPointFunctions)
```
> Can we have it so the functions are added to the existing map instead of overriding it?

Yes, that can work that too. I updated the `FactoryOptions` in my last commit https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39966/commits/cc848b6666d93b20314d7e5013d6ef71d790fbee . 

The new behaviour overrides the "default" functions on its first use of any `With*Functions` and every subsequent use merges the new functions to "current" functions map.

Now functions can be set like this :
```
NewFactoryWithOptions(
	WithLogFunctions(transformprocessor.DefaultLogFunctions()),
	WithLogFunctions([]ottl.Factory[ottllog.TransformContext]{FunctionOne, FunctionTwo, ...}),
	WithLogFunctions([]ottl.Factory[ottllog.TransformContext]{FunctionThree, FunctionFour, ...}),
)
```

evan-bradley replied:
My bad, I missed the variadic argument in the `With*Functions` functions. I think both of these are roughly equivalent, though I still have a slight preference for the new way of doing it.

@edmocosta @TylerHelmuth Please let us know if you feel differently about how this should look.

franciscovalentecastro committed:
[processor/transform] Fix unnecessary update.

evan-bradley commented on processor/transformprocessor/factory_test.go:
```diff
@@ -753,3 +779,237 @@ func TestFactoryCreateSpanEventProcessor(t *testing.T) {
+			statements: []common.ContextStatements{
+				{
+					Context:    common.ContextID("span"),
+					Statements: []string{`set(cache["attr"], TestSpanFunc())`},
```
Could you also have cases that show the default functions can be removed from the list?

evan-bradley reviewed (commented):
Thanks for your persistence through this review process @franciscovalentecastro. This looks good to me, just one request for a few additional test cases.

franciscovalentecastro committed:
[processor/transform] Add `api` to change_logs.

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

franciscovalentecastro committed:
[processor/transform] Update With*Functions comments, non-default functions log messages and fix typo.

franciscovalentecastro committed:
[process/transform] Update test to add cases without default functions.

franciscovalentecastro commented on processor/transformprocessor/factory.go:
```diff
@@ -22,34 +27,133 @@ import (
+// FactoryOption applies changes to transformProcessorFactory.
+type FactoryOption func(factory *transformProcessorFactory)
+
+// WithDataPointFunctions set ottl datapoint functions in resulting processor.
```
Done. I added the following function comments :
```
// WithLogFunctions will override the default ottl log context functions with the provided logFunctions in the resulting processor.
// Subsequent uses of WithLogFunctions will merge the provided logFunctions with the previously registered functions.
func WithLogFunctions(logFunctions []ottl.Factory[ottllog.TransformContext]) FactoryOption {
```

franciscovalentecastro committed:
[processor/transform] Update to uppercase OTTL.

franciscovalentecastro committed:
[processor/transform] Add log with fields to know which context registered non-default functions.

franciscovalentecastro committed:
[process/transform] Nit, remove `s` from traces.

evan-bradley committed:
Merge branch 'main' into fcovalente-with-add-ottl-funcs

evan-bradley committed:
Merge branch 'main' into fcovalente-with-add-ottl-funcs

franciscovalentecastro commented:
@evan-bradley @edmocosta @quentinmit Thank you 🙏  for all the reviews, comments and design guidance throughout this FR / PR !
